lib/ansible/module_utils/common/validator.py

```diff
@@ -0,0 +1,177 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2020 Ansible Project
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+from ansible.module_utils.common._collections_compat import (
+    KeysView,
+    Sequence,
+)
+
+from ansible.module_utils.common.parameters import (
+    get_unsupported_parameters,
+    get_type_validator,
+)
+
+from ansible.module_utils.parsing.convert_bool import BOOLEANS_FALSE, BOOLEANS_TRUE
+from ansible.module_utils.common.text.converters import to_native
+
+from ansible.module_utils.common.text.formatters import (
+    lenient_lowercase,
+)
+
+from ansible.module_utils.common.validation import (
+    check_required_arguments,
+    check_required_by,
+    check_required_if,
+    check_required_one_of,
+    check_required_together,
+)
+
+from ansible.module_utils.six import (
+    binary_type,
+    text_type,
+)
+
+
+class Validator():
+    """Argument spec validator class"""
+
+    def __init__(self, arg_spec, module_params):
+        self._options_context = None
+
+        self.arg_spec = arg_spec
+        self.module_params = module_params
+        self.error_messages = []
+        self.validated_params = {}
+
+    def _validate_argument_types(self):
+        """Validate module parameters and store them in self.validated_params"""
+
+        # TODO: Move to standalone function
+        # FIXME: Need to handle subelements
+
+        for k, v in self.arg_spec.items():
+            type = v.get('type')
+            if k not in self.module_params:
+                continue
+
+            value = self.module_params[k]
+            if value is None:
+                continue
+
+            type_checker, wanted_name = get_type_validator(type)
+
+            self.validated_params[k] = type_checker(value)
```

Comment:
In order to get _all_ the failures, I'm going to have to catch the TypeErrors here and return them all. Currently, this bails on the first validation failure. Something to keep in mind if I try to make this a standalone function.